### DISCUSSION TITLE
Octoprint

### DISCUSSION THREAD

**Discussion by aegean-odyssey** [aegean-odyssey/mpmd_marlin_1.1.x#50] [General]
> @drji3m writes:
>
> I installed on mynew delta.. Great work on firmware.
> 
> But im ready to return the delta since the usb serial is totally worthless.
> Most times octoprint wont connect, other times it fails while printing.
> 
> it gets worse. If i reboot the pi the printer hangs the pi and it cant reboot. Monoprice lives in a world of its own and cant make a simple USB interface work. The mainboard must go.
> 
> But myquestion to you is whether you think your firmware will install on a creality ender 3 maniboard-its a stm 32 processor. Im willing to try this swap before i give up on Monoprice.
>
> ender 3 pro stm32f103ret6

The attached image is not my "print farm", but these three machines run exclusively with OctoPrint on a Raspberry Pi and `mpmd_marlin_1.1.x` firmware. So maybe don't give up just yet. If you're using the latest release (119r15), there were changes to the USB code, so maybe I broke something. Trying the previous release (119r14) should shed some light on that possibility.

Also, I seem to recall that there was some kind of patch or add-on to get OctoPrint working with Monoprice MP Mini Delta. It is a work-around for a machine with the factory firmware -- it closed and re-opened the USB port to fix something with communication. This fix is ___not___ needed with `mpmd_marlin_1.1.x` firmware ~~and could interfere with smooth operation~~.

The controller board in the MP Mini Delta is based on a CPU from the _STM32F0_ family of controllers. It is not as powerful as the STM32F103 part and suffers from limited flash program space. It seems it's been dropped from the "supported" list of controller boards in the Marlin project with claims that it is not enough computer to run Marlin -- don't know what to say about that. On the other hand, though, I believe that the STM32F103 part is supported, so maybe using mainstream Marlin 2.0 for the ender 3 pro is a route to choose.  

![PrintFarm](https://user-images.githubusercontent.com/57586704/103606596-b7663f80-4f0e-11eb-9d6b-065a8c9f2bd3.jpg)

**Comment by PurpleHullPeas** [comment]
I used to have problems with Octoprint...until I got better hardware. It was really my fault for being cheap. A bad/cheap PSU (led blinking), SD Card (counterfeit crap), USB cable, and/or pi (Octoprint recommends at least a 3B) could lead to frequent drops. The connection fix plugin does not adversely affect MPMD Marlin 1.1.X once you're connected. Using M665 S120 also seems to help when printing certain geometries over USB.

**Comment by drji3m** [comment]
I'm using a pi4 2gb with full raspian and octoprint (also tried octopi,
same issues)

The connection problems occur with both the factory firmware and yours
which makes it seem to be in the delta usb-serial implementation. I may
also have a defective unit.

I've been at this for days.

What works is to power both printer and pi at the same time on power strip.
This almost always makes an operating connection.

But a print job won't complete without a comm failure at some point. I have
a cable with chokes for noise.

So I loaded your plugin for fast transfer to SD card. This fails utterly
with a broken connection. It also fails upload to sd without your plugin so
the interface can't handle fast data traffic.

If I reboot the pi the restart hangs the pi until the printer is unplugged.
This prevents remote reboot.

I also tried resetting the USB driver on the pi with no effect.

I guess the MPMD goes back to Amazon unless I replace the mboard  (if mine
is defective) or try to get the creality board to work the delta.

Im not at home so I'm not sure what version of your sw I'm running but
since is was downloaded last week its probably 14. I'll try the 15 today.

BTW, are they using a real chip for usb-serial or some AVR CPU?

Since the only complaint everyone else has is its hard to connect odds are
I have a lemon.

Jim V.

**Comment by aegean-odyssey** [comment]
@PurpleHullPeas, it's good to hear that the printer works for you with Octoprint -- helpful information. Is this the way you normally operate the printer, that is, printing directly from the host (not printing from the SD card)? 

I don't use Octoprint here, but I did a little testing wayback with Octoprint and Pronterface on a linux laptop. Octoprint in particular seemed pretty smart about adapting to the printer capabilities it finds when it sends an `M115` (firmware info) command.

On one or both of the hosts (Octoprint, Pronterface), though, I had to disable a "disconnect on error" setting to make operation more robust.

I logged quite a bit of time testing out the printer's custom `M990` command with Octoprint, Pronterface, picoterm (a serial terminal program), and from the linux command line. It seemed to work pretty well here and it is __MUCH__ faster than the Marlin upload. I thought speeding up file transfer would be useful, but sadly I don't think many folks use the feature.

@drji3m, I agree quality control is __not__ Monoprice's forté.  The company seems unconcerned about sending out defective machines, or machines with missing parts or misconfigured -- basically little regard for customers' satisfaction or frustration. No help from Monprice seems to be the implied warranty for the machine.

Anyway, I know that the printer with `mpmd_marlin_1.1.x` firmware can work better than your experience tells you. Can you provide any logs of a faulty Octoprint session? Maybe we can figure something out.

**Comment by PurpleHullPeas** [comment]
@aegean-odyssey Yes, straight from Octoprint has been my primary method of printing on my two MPMDs for a very long time. I have frequently done prints spanning 6-24 hours with fine detail and small nozzles. Your SD Card upload function has also worked quite well for the couple of times I have tried it (very recently). 

I also agree with @drji3m and @aegean-odyssey 's sentiments regarding crap quality control. There seems to be no consistency regarding machine alignment, firmware setup/settings (e.g., M502 will sometimes bork a stock printer's settings), machine damage, SD card quality, etc. Based on what I've seen in the Facebook Group, I'm pretty sure Monoprice's refurbishing process for "Open Box" machines is simply changing the shipping label. At least a couple of times, people showed up with their "brand new" MPMD that was clearly used (scratches, ripped stickers, aftermarket mods, etc.). 

I mention all of that to say, if @drji3m is reporting that he has already checked all of the potential problem-areas I've mentioned, then I would agree that his particular MPMD is a lemon.

**Comment by drji3m** [comment]
The bad delta is on its way back to Amazon after I removed all my upgrades
and put back the factory firmware.

I'm fortunate to have a local Microcenter so I picked up a new one.

It like magic. All the problems evaporated. It works perfectly and is
actually quieter than the first. So the conclusion is the first was broken
in mysterious ways. Not our problem anymore.

I think I'll wait a couple of days before installing your firmware to
isolate any problems which might show up.

I had replaced the bed sticker with a magnetic flex surface on the bad one.
So I thought I would just swap build plates with the new one. Little did I
know that they changed the thermistor connector between units.

So it seems Monoprice can't make two alike and works outside the industrial
revolution. Just like adding Mcode to reverse motors  instead of
controlling motor wiring.

I plan to use Octoprint but also use your fast upload to SD since printing
from SD seems more robust. Yet I print for hours with Octoprint  on my 3
Wanhoe I3s with no problems.

It was my plan to give a MPMD to my grandaughter for her 13 birthday but
I've got to make it foolproof first.

I appreciate this conversation and will keep you posted on results with v15
of your firmware.

**Comment by aegean-odyssey** [comment]
BTW,  @tclupper has a very nice looking and functional Pi mounting bracket for the MPMD in his `MPMDextras` repository. It's called [__PiHolder__](https://github.com/tclupper/MPMDextras/)
